### PR TITLE
Stop processing split queries on first error.

### PR DESCRIPTION
**Pull request by jeschkies**
**What this PR does / why we need it**:
The `TestMetricsTripperware` became flaky after #10688. The race condition in the splitter has been there before but the change exposed it. When an error occurs, e.g. when query would be too large, the splitter would return. However, there was a small time window when the loop could still be fed a new request. #10688 just amplified this effect because it removed the serialization of requests and responses.

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory.

### CODE REVIEW

**Review by kavirajk — APPROVED** [review]
Nice 👍

File: pkg/querier/queryrange/roundtrip_test.go [hunk 1]
```diff
@@ -207,7 +207,7 @@ func TestMetricsTripperware(t *testing.T) {
 	queryCount, queryHandler = counter()
 	h = getQueryAndStatsHandler(queryHandler, statsHandler)
 	_, err = tpw.Wrap(h).Do(ctx, lreq)
-	require.Error(t, err)
+	require.ErrorContains(t, err, "query too large to execute on a single querie")
```
Comment (dannykopping) [inline pkg/querier/queryrange/roundtrip_test.go:210]: Tiniest of nits, triggered my OCD 😛 feel free to ignore

```suggestion
	require.ErrorContains(t, err, "query too large to execute on a single querier")
```

**Review by dannykopping — APPROVED** [review]
Was super pairing up with you to find and fix this 💪 great work!

File: pkg/querier/queryrange/roundtrip_test.go (hunk 1 above)
Comment (jeschkies) [reply pkg/querier/queryrange/roundtrip_test.go:210]: :joy:

**Pull request merged**